google/cloud/storage/transfer_manager.py

```diff
@@ -844,9 +855,208 @@ def download_chunks_concurrently(
+        detail of the Client object, especially if the Client was modified after
+        its initial creation or if `Client._http` was modified in any way.
+
+        THREAD worker types are observed to be relatively efficient for
+        operations with many small files, but not for operations with large
+        files. PROCESS workers are recommended for large file operations.
+
+    :type max_workers: int
+    :param max_workers:
+        The maximum number of workers to create to handle the workload.
+
+        With PROCESS workers, a larger number of workers will consume more
+        system resources (memory and CPU) at once.
+
+        How many workers is optimal depends heavily on the specific use case,
+        and the default is a conservative number that should work okay in most
+        cases without consuming excessive resources.
+
+    :type checksum: str
+    :param checksum:
+        (Optional) The checksum scheme to use: either 'md5', 'crc32c' or None.
+        Each individual part is checksummed. At present, the selected checksum
+        rule is only applied to parts and a separate checksum of the entire
+        resulting blob is not computed. Please compute and compare the checksum
+        of the file to the resulting blob separately if needed, using the
+        'crc32c' algorithm as per the XML MPU documentation.
+
+    :type timeout: float or tuple
+    :param timeout:
+        (Optional) The amount of time, in seconds, to wait
+        for the server response.  See: :ref:`configuring_timeouts`
+
+    :raises: :exc:`concurrent.futures.TimeoutError` if deadline is exceeded.
+    """
+
+    # TODO:
+    # make base url creation more robust than string addition - consider also using that for query strings in resumable media
+    # metadata
+    # other special purpose header stuff like encryption
+    # figure out custom client stuff
+
+    bucket = blob.bucket
+    client = blob.client
+    transport = blob._get_transport(client)
+
+    hostname = _get_host_name(client._connection)
+    url = hostname + "/" + bucket.name + "/" + blob.name  # FIXME: make this nicer
+    content_type = blob._get_content_type(content_type, filename=filename)
+
+    container = XMLMPUContainer(url, filename)
+    container.initiate(transport=transport, content_type=content_type)
+    upload_id = container.upload_id
+
+    size = os.path.getsize(filename)
+    num_of_parts = -(size // -chunk_size)
+
+    pool_class, needs_pickling = _get_pool_class_and_requirements(worker_type)
+    # Pickle the blob ahead of time (just once, not once per chunk) if needed.
+    maybe_pickled_client = _pickle_client(client) if needs_pickling else client
+
+    futures = []
+
+    with pool_class(max_workers=max_workers) as executor:
+
+        for part_number in range(1, num_of_parts + 1):
+            start = (part_number - 1) * chunk_size
+            end = min(part_number * chunk_size, size)
```

Comment:
Could you remind me again if the byte ranges start/end are inclusive? 

I'm wondering for the intermediate parts, should this instead be `end = min(part_number * chunk_size - 1, size)`?

Reply:
It's zero-indexed and exclusive, which means that if part 1 is 0-8 and part 2 is 8-16, then part 1 will have bytes 0, 1, 2, 3, 4, 5, 6, 7 and part 2 will have 8, 9, 10, 11, 12, 13, 14, 15. Both parts have zero bytes and the total is 16. Byte 16 doesn't exist. So this math should be accurate - I'll be sure to triple-check by actually using the method on some large files and checksumming them, though.